[#10] Feat: ReportedImageList UI 구현

HY219 opened the pull request:
## 📝 작업 내용
- 신고된 이미지의 목록을 생성하고, 각 이미지의 신고 횟수를 구하는 로직 구현
  - 중복된 이미지 아이디 없이 신고된 이미지 리스트 출력
  - 신고된 이미지의 신고 횟수 출력


- 가장 최근에 신고된 날짜로 Date 출력

-  ReportedDetailButton 클릭 시 imageId 전달


## 📷 스크린샷 (선택)
<img width="528" alt="image" src="https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/e77cfb02-b91f-4766-a7eb-4edfe4f786d4">

## 💬 리뷰 요구사항(선택)
- `type`, `변수명`, `폴더명`을 올바르게 작성했는지 궁금합니다!

- 데이터가 많아질 경우 지금 로직(`중복 이미지 아이디`를 출력하지 않고, `신고 횟수`를 얻는) 시간이 많이 소요될까요?

- `pages폴더` 없이 `routes폴더`내에 page를 이루는 컴포넌트들이 import되는 것이 괜찮은가요?

## 📍 기타 (선택)

- 예시 테이블을 참고해서 신고된 게시물 리스트 목록 출력하는 로직을 작성했습니다. 
데이터가 많아질 경우 지금 로직이 시간이 많이 소요된다면 DB 구조 변경이 가능한지 BE에 여쭤보면 좋을 것 같습니다:)

  <img width="343" alt="image" src="https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/9882cece-7fad-4b35-afeb-ebc5815fb9fc">

- `TODO:` prettier에러로 추후 사용자 지정 css 추가 필요

- 추가로, `routes폴더`에 `page`를 작성하는 것이 어색하게 느껴져서 
아래 사진의 방식처럼, `pages폴더`를 만들어 코드를 작성한 후, 그 컴포넌트를 `routes폴더`에 가져오는 것은 어떤가요?
더 좋은 방식이 있다면 여쭤보고 싶습니다:)
  <img width="858" alt="image" src="https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/2e992d8b-a05c-403e-8360-1b5014793cf4">


close #10

HY219 committed:
Feat: ReportedDetailList 구현

admin 페이지 구현
중복 없이 신고된 이미지의 신고 횟수 출력
prettier에러로 추후 사용자 지정 css 추가 필요

HY219 committed:
Feat: ReportedDetailButton 클릭 시 imageId 전달

HY219 committed:
Fix: tailwind.config.js의 기본 색상 적용 코드 변경 및 사용자 지정 css 추가

button과 text에 사용자 지정 css 적용

yjc2021 commented on src/components/common/admin/ReportedDetailButton.tsx:
```diff
@@ -0,0 +1,19 @@
+interface ReportedDetailButton {
```
네이밍 Props로 바꿔주시면 좋을 것 같습니다 :)
(convention)

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+import ReportedDetailButton from "./ReportedDetailButton";
+
+interface ReportedImageList {
```
인터페이스명이랑 컴포넌트 이름이랑 중복이네요! 둘이 구별해줘야 할 것 같습니다

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+}
+
+const ReportedImageList = () => {
+  const reportData = [
```
P5
더미 테이터는 컴포넌트 렌더마다 다시 만들 필요 없어서 컴포넌트 밖에다 정의하셔도 될 것 같습니다

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+    },
+  ];
+
+  const reportedImage: ReportedImageList[] = [];
```
변수 네이밍이 조금 애매한 것 같아요! 리스트를 담는 변수니까 복수가 되어야 할 것 같네요.
`reportedImages` 처럼요 !

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+
+  const reportedImage: ReportedImageList[] = [];
+
+  reportData.forEach((item) => {
```
서버 쪽 데이터를 클라이언트 네이밍에 맞게 바꾸는 과정인거죠??
함수로 만들면 함수 네이밍이 있어서 코드 가독성이 높아질 것 같아요.
그리고 utils 폴더에 따로 분리해도 될 것 같아요

yjc2021 replied:
함수로 만든다면 함수가 parsing된 데이터 리스트를 반환하면 될 것 같네요 :)
```ts
const reportedImage = parseImageList(repotData)

// 네이밍은 대충 한 거에요!!
```

yjc2021 replied:
재민님 리뷰 보니까 재사용 되지 않는 로직은 따로 파일 분리를 할 필요는 없을 것 같더라구요! 그래서 파일 안에서 함수로만 분리하는 게 좋을 것 같아요 :)

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+    const imageId = item.image_id;
+    const createdAt = item.created_at;
+    const existingItem = reportedImage.find((obj) => obj.imageId === imageId);
+    existingItem
```
흐음 신고횟수는 클라이언트에서 계산할 필요가 있을까요? 아마 BE에서 내려줄 때 값도 같이 주지 않을까 싶은데 함 물어보시죠 👍

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+          {reportedImage.map((item, index) => {
+            return (
+              <tr key={index} className="border-b-1 border-card last:border-0">
+                <td className="text-center text-text-primary">{item.createdAt.slice(0, 10)}</td>
```
`item.createdAt.slice(0,10)` 도 어떤 동작을 하는지 한 눈에 파악이 어려운 것 같아요. 명시적으로 함수로 분리하는 건 어떨까요?

yjc2021 commented on src/routes/admin/index.tsx:
```diff
@@ -1,7 +1,12 @@
 const Admin = () => {
-  return <div>관리자 페이지</div>;
+  return (
+    <div>
```
페이지의 layout 스타일링하시는 걸 깜빡하신 것 같아요!
페이지 기본 레이아웃은 모든 페이지에서 통일되면 좋을 것 같아서 이건 이슈로 추가하고 다음에 구현해봐요 :)

HY219 replied:
공통으로 사용될 레이아웃이 나오면, 이후에 이슈 생성해서 작업하면 좋을 것 같아요:)

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+      ? (existingItem.reportCount += 1)
+      : reportedImage.push({ imageId, reportCount: 1, createdAt });
+  });
+
```
스타일에 라이트/다크 테마 고려가 안 된 것 같아요. 와이어프레임 참고하셔서 tailwind config에 등록된 색상으로 부탁드립니다! 
색상 네이밍 정책은 빠른 시일 내에 마지막으로 한번 개선하겠습니다. 마이그레이션 작업은 제가 이슈로 파서 기존 파일들도 다 적용할테니 일단은 현재 등록된 색상 토큰 사용해주세요

HY219 replied:
혹시 아직 코드에 등록된 색상 토큰이 적용되지 않은 부분이 있을까요??

yjc2021 reviewed (commented):
혜연님 수고하셨습니다 :)

yjc2021 commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+        <tbody>
+          {reportedImage.map((item, index) => {
+            return (
+              <tr key={index} className="border-b-1 border-card last:border-0">
```
map을 돌 때 각각의 item의 key는 저희 컨벤션대로 수정해주세요!

choi-ik commented on src/components/common/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,73 @@
+    existingItem
+      ? (existingItem.reportCount += 1)
+      : reportedImage.push({ imageId, reportCount: 1, createdAt });
+  });
```
음 저는 이 부분이 여기서 왜 필요한지 잘 모르겠습니다! 서버에서 날짜, id, 신고 횟수 userId등 모든 데이터를 내려줄텐데 한번 더 이렇게 데이터를 핸들링 해야 할 이유가 있을까요?? 이 부분을 차라리 테이블 바이 부분에서 진행 해주는게 좋을 것 같고, 거기서도 뭔가 계산식을 넣는게 아니라 넘겨준 데이터를 반복문을 돌며 뿌려주기만 하면 될 것 같습니다! 👍🏻👍🏻

choi-ik reviewed (commented):
헤연님 구현 하시느라 고생 많으셨습니다 ㅎㅎ

choi-ik reviewed (commented):
헤연님 구현 하시느라 고생 많으셨습니다 ㅎㅎ

choi-ik reviewed (commented):
헤연님 구현 하시느라 고생 많으셨습니다 ㅎㅎ 수정이 필요하다 느껴지는 부분은 재민님과 용재님이 다 해주신 것 같아서 저는 별로 남길 말이 없네요 ㅎㅎ 다만 admin이 common 폴더에 들어가는게 맞는 건지는 좀 의문입니다! 뭔가 공통적으로 쓰이지는 않는 것 같거든요 ㅎㅎ 이 부분도 고민해보시면 좋을 것 같습니다! 고생하셨습니다 🙇‍♂️🙇‍♂️

HY219 committed:
Rename: admin디렉터리 구조 변경 및 명확한 변수명으로 변경

admin디렉터리를 components/로 위치 변경
reportData -> reportLists
reportedImage -> reportedImages
obj -> reportedImageObj

hyeonjinan096 reviewed (commented):
혜연님 수고하셨습니다!!👍

HY219 committed:
Refactor: 상세보기 버튼 컴포넌트를 제거하고 인라인 코드로 변경

ReportedDetailButton 컴포넌트 제거
상세보기 버튼 컴포넌트를 제거하고 부모 컴포넌트 안에서 직접 버튼 코드를 인라인으로 작성
코드 구조 단순화하고 불필요한 컴포넌트 부하를 줄인다.

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,84 @@
+interface ReportedImageList {
```
이름을 Props 타입으로 하기로 했어요!

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,84 @@
+}
+
+const ReportedImageList = () => {
+  const reportLists = [
```
`reportLists`면 이런 배열 리스트가 여러 개 있다는 의미가 돼서 그냥 `reports`라고 해주시는게 더 좋을 것 같아요

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,84 @@
+    },
+  ];
+
+  const reportedImages: ReportedImageList[] = [];
+
+  reportLists.forEach((reportedImage) => {
+    const imageId = reportedImage.image_id;
+    const createdAt = reportedImage.created_at;
+    const existingItem = reportedImages.find(
+      (reportedImageObj) => reportedImageObj.imageId === imageId,
+    );
+    existingItem
+      ? (existingItem.reportCount += 1)
+      : reportedImages.push({ imageId, reportCount: 1, createdAt });
+  });
```
이 부분은 사실 없어도 될 것 같아요. `reportedImages`를 저희가 가지고 있는 것도 아니고 따로 서버 요청을 할 것도 아니라서요. 그냥 서버 응답에 신고 횟수를 같이 내려달라고 하는게 나을 것 같습ㄴ디ㅏ.

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,84 @@
+                <td className="text-center text-text-primary">
+                  <button
+                    className="btn btn-neutral btn-sm text-xs"
+                    onClick={() => handleClickOpen(reportedImage.imageId)}
```
이 부분은 아래처럼 커링 사용하시면 해결됩니다! handleClickOpen은 잘못된 이름 같아요. 저희 컨벤션이 `handle + event + 대상`인데 대상 부분에 동사가 와버려서 의미를 알기가 어려워진 것 같습니다. 상세보기 버튼이니까 `handleClickDetailViewButton`은 어떨까요?

```js
const handleClickOpen = (id: string) => () => {};

onClick={handleClickOpen(imageId)}
```

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,84 @@
+                <td className="text-center text-text-primary">
+                  {reportedImage.createdAt.slice(0, 10)}
+                </td>
+                <td className="text-center text-text-primary">{reportedImage.imageId}</td>
+                <td className="text-center text-text-primary">{reportedImage.reportCount}</td>
```
이거 `reportedImage.`도 중복 코드의 일종이에요! 디스트럭처링해주세요~

HY219 committed:
Cleanup: 이미지 신고 횟수를 구하는 로직 삭제

BE에서 값을 넘겨받으면 되므로 불필요

임시 데이터에 reportCount 추가

HY219 committed:
Rename: 변수명 및 talbe 요소의 key값 변경

reportLists -> reports
handleClickOpen -> handleClickDetailViewButton
reportedImage -> reportImage
역할이 명확하게 변수명 변경
고유한 값을 가지도록 컨벤션에 맞게 key값 변경

HY219 committed:
Refactor: 중복 코드를 디스트럭처링 처리 및 가공한 임시데이터를 변수에 저장

reportImage. 중복코드를 제거
createDate 변수에 신고된 이미지가 등록된 Date를 저장

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,68 @@
+          </tr>
+        </thead>
+        <tbody>
+          {reports.map((reportImage, index) => {
```
여기서 바로 distructuring 하시면 돼요! 
```js
{reports.map(({createdDate, imageId, reportCount}, index) => {}
```

HY219 replied:
오 감사합니다!! 수정하겠습니다~!

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
```diff
@@ -0,0 +1,68 @@
+                <td className="text-center text-text-primary">
+                  <button
+                    className="btn btn-neutral btn-sm text-xs"
+                    onClick={() => handleClickDetailViewButton(imageId)}
```
[이거](https://github.com/zzalmyu/zzalmyu-frontend/pull/27#discussion_r1490307873) 반영 안된거 같습니다ㅠ

HY219 replied:
앗 그러네요 반영했습니다!

thalassophilia commented on src/components/admin/ReportedImageList.tsx:
지금 봤는데 `components/Admin`이 맞을 것 같네요!

HY219 committed:
Refactor: 커링을 사용해 인라인 함수 제거

커링을 사용해, onClick에서 인라인 함수로 선언한 코드 제거
& 디스트럭처링 할당에서 불필요한 변수를 제거하여 수정
key값 변경이 반영이 안되어 명시적으로 재변경

thalassophilia commented:
배포 실패 에러 메시지입니다~
<img width="1279" alt="image" src="https://github.com/zzalmyu/zzalmyu-frontend/assets/61978339/3a19fc7e-6890-44da-8236-8bf060caf47a">

HY219 committed:
Fix: git 대소문자 구분 이슈 해결

git-mv를 이용해 dir명의 대소문자 구분하여 변경

HY219 committed:
Merge branch 'main' into #10/feature/report-list-page-ui

yjc2021 reviewed (approved):
LGTM! 고생하셨습니다
